### PR TITLE
Allow dropins for udev.conf

### LINKED ISSUE #30460: Teach systemd-udevd to support dropin files for `udev.conf`
Currently `udev.conf` doesn't support dropins, which seems to be an annoying exception compared to our other stuff. It would be nice if the behavior was consistent, even though this particular config file has only a couple of lines.

Originally mentioned in https://github.com/systemd/systemd/pull/30387#issue-2033069437.

### PR DESCRIPTION
**Pull request by dtardon**
Fixes #30460 #19488

### CODE REVIEW

File: test/units/testsuite-17.03.sh [hunk 1]
```diff
@@ -8,7 +8,7 @@ KILL_PID=
 
 setup() {
     mkdir -p "${TEST_RULE%/*}"
-    [[ -e /etc/udev/udev.conf ]] && cp -f /etc/udev/udev.conf /etc/udev/udev.conf.bak
+    mkdir -p /etc/udev/udev.conf.d
```
Comment (yuwata) [inline test/units/testsuite-17.03.sh]: Please use `/run` now.

File: man/udev.conf.xml [hunk 2]
```diff
@@ -18,23 +18,35 @@
 
   <refnamediv>
     <refname>udev.conf</refname>
+    <refname>udev.conf.d</refname>
```
Comment (yuwata) [inline man/udev.conf.xml:21]: Have you run update-man-rule ?

**Review by yuwata — APPROVED** [review]

File: src/shared/conf-parser.c [hunk 3]
```diff
@@ -598,10 +598,11 @@ static int config_parse_many_files(
         return 0;
 }
 
-/* Parse one main config file located in /etc/systemd and its drop-ins, which is what all systemd daemons
+/* Parse one main config file located in /etc/$domain and its drop-ins, which is what all systemd daemons
  * do. */
-int config_parse_config_file(
+int config_parse_config_file_full(
                 const char *conf_file,
+                const char *domain,
```
Comment (poettering) [inline src/shared/conf-parser.c]: maybe call this pkgdir?

at least in autoconf that's what it was known by pkgconfdir, pkglibdir, …

File: src/udev/udevd.c [hunk 4]
```diff
@@ -65,70 +66,37 @@ static int listen_fds(int *ret_ctrl, int *ret_netlink) {
...
-                           "udev_log", &log_val,
-                           "children_max", &children_max,
-                           "exec_delay", &exec_delay,
-                           "event_timeout", &event_timeout,
-                           "resolve_names", &resolve_names,
-                           "timeout_signal", &timeout_signal);
+        r = config_parse_config_file_full(
+                        "udev.conf",
+                        "udev",
+                        NULL,
```
Comment (poettering) [inline src/udev/udevd.c]: as per current coding style, please prefix opaque parameters like this with their name, i.e. change this line to `/* sections= */ NULL,`

File: src/udev/udevd.c [hunk 5]
```diff
@@ -65,70 +66,37 @@ static int listen_fds(int *ret_ctrl, int *ret_netlink) {
...
+                        config_item_table_lookup,
+                        config_table,
+                        CONFIG_PARSE_WARN,
+                        NULL);
```
Comment (poettering) [inline src/udev/udevd.c]: similar

File: src/shared/udev-util.c [hunk 6]
```diff
@@ -45,20 +46,28 @@ int udev_set_max_log_level(char *str) {
...
-        r = parse_env_file(NULL, "/etc/udev/udev.conf",
-                           "udev_log", &log_val);
+        r = config_parse_config_file_full(
+                        "udev.conf",
+                        "udev",
+                        NULL,
+                        config_item_table_lookup,
+                        config_table,
+                        CONFIG_PARSE_WARN,
+                        NULL);
```
Comment (poettering) [inline src/shared/udev-util.c]: as above

File: test/units/testsuite-17.03.sh (hunk 1 above)
Comment (dtardon) [reply test/units/testsuite-17.03.sh]: Eh. I thought I did...

File: man/udev.conf.xml (hunk 2 above)
Comment (dtardon) [reply man/udev.conf.xml:21]: I have now.

File: src/shared/conf-parser.c (hunk 3 above)
Comment (dtardon) [reply src/shared/conf-parser.c]: Fine with me...

File: src/shared/conf-parser.c [hunk 7]
```diff
@@ -626,7 +628,7 @@ int config_parse_config_file(
         STRV_FOREACH(p, conf_paths) {
                 char *d;
 
-                d = strjoin(*p, "systemd/", conf_file, ".d");
+                d = strjoin(*p, domain, "/", conf_file, ".d");
```
Comment (mrc0mmand) [inline src/shared/conf-parser.c]: domain -> pkgdir

Comment (dtardon) [reply src/shared/conf-parser.c]: Eh. How did that even compile?!

File: src/udev/udevd.c [hunk 8]
```diff
@@ -65,70 +66,37 @@ static int listen_fds(int *ret_ctrl, int *ret_netlink) {
         return 0;
 }
 
+static DEFINE_CONFIG_PARSE_ENUM(config_parse_resolve_name_timing, resolve_name_timing, ResolveNameTiming, "Failed to parse resolve name timing");
+
 static int manager_parse_udev_config(Manager *manager) {
-        _cleanup_free_ char *log_val = NULL, *children_max = NULL, *exec_delay = NULL,
-                *event_timeout = NULL, *resolve_names = NULL, *timeout_signal = NULL;
+        int log_val = -1;
+        Manager *m = ASSERT_PTR(manager); /* A hack to check the ptr before using it in the table */
```
Comment (YHNdnzj) [inline src/udev/udevd.c]: Why not use normal assertion?

Comment (dtardon) [reply src/udev/udevd.c]: > Why not use normal assertion?

We don't mix `assert`s and declarations.

I could have used `ASSERT_PTR()` just in the first line of the dispatch table (or even in all lines), but the current variant with an intermediate variable appeared slightly nicer to me.

Comment (YHNdnzj) [reply src/udev/udevd.c]: > We don't mix asserts and declarations.

You can define the table after variable declarations and assertions, as with `shared/sleep-config.c`.

Comment (yuwata) [reply src/udev/udevd.c]: Fixed.

**Pull request merged**